### PR TITLE
EMA based statistically adaptive thread pool design

### PR DESCRIPTION
**Pull request by vertexclique**
This is an adaptive thread pool implementation with exponential moving averages(where it doesn't rely on idle time).

Here are the benchmarks:

**This new statistical-EMA-approach:**
- CPU utilization ~97%

```
running 1 test
test blocking ... bench:  39,116,408 ns/iter (+/- 5,686,213)
```

**Current approach**
- CPU utilization ~88%

```
running 1 test
test blocking ... bench: 254,204,382 ns/iter (+/- 214,099,973)
```

This approach enables 5.5x speedup.

On top of that, this PR solves #65 intrinsically 
Just because it is not nice to manually include modules at the benchmark level.

## Documentation
blocking doc comment

### CODE REVIEW

**Comment by yoshuawuyts** [comment]
I'm in no means qualified to review this code, but the results are looking very cool! awesome work @vertexclique! :tada:

File: src/task/blocking.rs [hunk 1]
```diff
@@ -73,25 +96,49 @@ fn maybe_create_another_blocking_thread() {
...
+    FREQUENCY.fetch_add(1, Ordering::Relaxed);
+
+    // Calculate the amount of threads needed to spin up
+    // then retry sending while blocking. It doesn't spin if
+    // expected pool size is above the MAX_THREADS (which is a
+    // case won't happen)
+    let pool_size = EXPECTED_POOL_SIZE.load(Ordering::Relaxed);
+    let current_pool_size = CURRENT_POOL_SIZE.load(Ordering::Relaxed);
+    if pool_size > current_pool_size && pool_size <= MAX_THREADS {
+        let needed = pool_size - current_pool_size;
```
Comment (spacejam) [inline src/task/blocking.rs]: maybe a saturating subtraction for defense in depth?

Comment (vertexclique) [reply src/task/blocking.rs]: Was doing that rn. Just in place comment

File: src/task/blocking.rs [hunk 2]
```diff
@@ -47,18 +58,30 @@ lazy_static! {
...
-    // even if it's random.
-    let workers = DYNAMIC_THREAD_COUNT.load(Ordering::Relaxed);
-    if workers >= MAX_THREADS {
-        return;
+fn calculate_dispatch_frequency() {
+    // Calculate current message processing rate here
+    let previous_freq = FREQUENCY.fetch_sub(1, Ordering::Relaxed);
+    let avr_freq = AVR_FREQUENCY.load(Ordering::Relaxed);
+    let current_pool_size = CURRENT_POOL_SIZE.load(Ordering::Relaxed);
+    let frequency = (avr_freq as f64 + previous_freq as f64 / current_pool_size as f64) as u64;
```
Comment (spacejam) [inline src/task/blocking.rs]: Should the avr and previous freq's be collectively divided by the current pool size? It's not totally clear to me how this avoids creating a feedback loop.

Comment (vertexclique) [reply src/task/blocking.rs]: Actually, that holds the current one. I made it more clear.

Comment (vertexclique) [reply src/task/blocking.rs]: That's how moving average works in general. It won't create a feedback loop since finished threads will decrease the needed count and that saturating thread join will join them over time. It won't spawn new threads, because current pool size is controlling thread creation.

File: src/task/blocking.rs [hunk 3]
```diff
@@ -73,25 +100,53 @@ fn maybe_create_another_blocking_thread() {
...
             let wait_limit = Duration::from_millis(1000 + rand_sleep_ms);
 
-            DYNAMIC_THREAD_COUNT.fetch_add(1, Ordering::Relaxed);
+            CURRENT_POOL_SIZE.fetch_add(1, Ordering::Relaxed);
             while let Ok(task) = POOL.receiver.recv_timeout(wait_limit) {
                 abort_on_panic(|| task.run());
+                calculate_dispatch_frequency();
             }
-            DYNAMIC_THREAD_COUNT.fetch_sub(1, Ordering::Relaxed);
+            CURRENT_POOL_SIZE.fetch_sub(1, Ordering::Relaxed);
```
Comment (spacejam) [inline src/task/blocking.rs]: This can't be a Relaxed atomic if we want to guarantee a happens-before relationship with the addition, which is important to guarantee that we never divide by zero in the ema calculation above

Comment (spacejam) [reply src/task/blocking.rs]: (can also just add 1 to the divisor in the calculation above to enforce this invariant at the dependency site)

Comment (spacejam) [reply src/task/blocking.rs]: but atomics will wrap, so we probably want to use a proper barrier here anyway for defense against wrap-intolerance in other calculation areas

Comment (spacejam) [reply src/task/blocking.rs]: Ordering::Release for writes generally and Ordering::Acquire for all loads I think is OK on this code, but honestly due to the timescales involved in blocking codepaths it's fine to use a SeqCst to think even less about wrap issues.

Comment (vertexclique) [reply src/task/blocking.rs]: I am going to use SeqCst in both.

**Comment by spacejam** [comment]
Thanks for continuing to boil this down! These are really nice results that address the main runaway thread growth issue in the naive original implementation.

I'm working through some more correctness proofs on paper to try to find more ways to break the large number of Relaxed atomic accesses that are added in this.

**Comment by vertexclique** [comment]
Thanks, Tyler. Very nice to hear that. If you are doing FV please share with me too.

**Comment by spacejam** [comment]
Hehehe the reason I'm using paper for a couple things is because I'm a little uncertain about translating Relaxed stuff into TLA (maybe using sets of all reorderable code and permuting them?)

Basically, anywhere a completely random value impacts runtime correctness is the only thing I'm looking for and desk-checking backwards from.

**Comment by vertexclique** [comment]
More to improve (aka TO-DO):
* Adaptive scale down (we can do this sometime later, it is harder and need more sync points than above)
* Compartmentalize the pool creation into so-called `execution context`'s so they satisfy the needs in the issue #79 . This PR also helps that in an indirect way. (Atomics can be moved for that mentioned issue, no worries)

**Comment by spacejam** [comment]
If more synchronization needs to occur, it probably isn't a bad time to take a step back from the 100% atomics-based approach and either use ROWEX when an adaptation is required (spawning readers work lock-free, writers need to claim an exclusive mutex) or just using a Mutex so that the implementation is obviously correct. Thinking about all of these atomic accesses is quite labor intensive and error-prone.

**Comment by vertexclique** [comment]
I think todos can go to other PRs. For Mutexes (if I understand you correctly you are talking about the scaling down part) it can be useful, but I don't think they should be used for this code since there is no critical section code here, I don't like abusing lock-exclusion in places which are not needed. (But I am not fully against that).

I am rooting for this PR to take it like this for blocking pool like how it is rn. Then we will think on that later in the process of extraction (which will also be for async part, I think).

**Comment by spacejam** [comment]
I guess what I'm mostly getting at is that there are some potentially very subtle information flows that can happen when this math happens to be operating in an entirely concurrent way on global variables via Relaxed atomics (or even SeqCst atomics), and this is expensive overall to reason about as a human (it's also expensive/unreasonable for the formal verification tooling I know about to provide meaningful validations of).

**Comment by spacejam** [comment]
There is almost never a real need for exclusion, but it makes PR's like this understandable without staring at it for hours :P

File: src/task/blocking.rs [hunk 4]
```diff
@@ -47,18 +58,34 @@ lazy_static! {
...
-    // it's just a heuristic, and would not lose correctness
-    // even if it's random.
-    let workers = DYNAMIC_THREAD_COUNT.load(Ordering::Relaxed);
-    if workers >= MAX_THREADS {
-        return;
+fn calculate_dispatch_frequency() {
+    // Calculate current message processing rate here
+    let current_freq = FREQUENCY.fetch_sub(1, Ordering::Relaxed);
+    let avr_freq = AVR_FREQUENCY.load(Ordering::Relaxed);
+    let current_pool_size = CURRENT_POOL_SIZE.load(Ordering::Relaxed);
```
Comment (spacejam) [inline src/task/blocking.rs]: For defense in depth and Relaxed raciness protection:
`std::cmp::max(LOW_WATER_MARK, CURRENT_POOL_SIZE.load(Ordering::Relaxed)`

File: src/task/blocking.rs [hunk 5]
```diff
@@ -47,18 +58,34 @@ lazy_static! {
...
-fn maybe_create_another_blocking_thread() {
-    // We use a `Relaxed` atomic operation because
-    // it's just a heuristic, and would not lose correctness
-    // even if it's random.
-    let workers = DYNAMIC_THREAD_COUNT.load(Ordering::Relaxed);
-    if workers >= MAX_THREADS {
-        return;
+fn calculate_dispatch_frequency() {
+    // Calculate current message processing rate here
+    let current_freq = FREQUENCY.fetch_sub(1, Ordering::Relaxed);
```
Comment (spacejam) [inline src/task/blocking.rs]: This only happens to be properly serialized due to the SeqCst barrier placed on the CURRENT_POOL_SIZE load in schedule(), but if that changes this is vulnerable to an underwrap. Please make it at least `Ordering::Acquire` for defense.

File: src/task/blocking.rs [hunk 6]
```diff
@@ -73,25 +100,53 @@ fn maybe_create_another_blocking_thread() {
...
+// based on the previous statistics without relying on
+// if there is not a thread ready to accept the work or not.
 fn schedule(t: async_task::Task<()>) {
+    // Add up for every incoming task schedule
+    FREQUENCY.fetch_add(1, Ordering::Relaxed);
+
+    // Calculate the amount of threads needed to spin up
+    // then retry sending while blocking. It doesn't spin if
+    // expected pool size is above the MAX_THREADS (which is a
+    // case won't happen)
```
Comment (spacejam) [inline src/task/blocking.rs]: This case comes really close to being able to happen though! Without the SeqCst barrier on CURRENT_POOL_SIZE below, AVR_FREQUENCY and FREQUENCY would be able to underwrap in calculate_dispatch_frequency. To make it really clear that this won't happen, proper barriers should be set on the atomic accesses on FREQUENCY + AVR_FREQUENCY as well.

File: src/task/blocking.rs [hunk 7]
```diff
@@ -73,25 +100,53 @@ fn maybe_create_another_blocking_thread() {
...
 
 // Enqueues work, attempting to send to the threadpool in a
-// nonblocking way and spinning up another worker thread if
-// there is not a thread ready to accept the work.
+// nonblocking way and spinning up needed amount of threads
+// based on the previous statistics without relying on
+// if there is not a thread ready to accept the work or not.
 fn schedule(t: async_task::Task<()>) {
+    // Add up for every incoming task schedule
+    FREQUENCY.fetch_add(1, Ordering::Relaxed);
```
Comment (spacejam) [inline src/task/blocking.rs]: Please use an `Ordering::Acquire` or stronger here

File: src/task/blocking.rs [hunk 8]
```diff
@@ -73,25 +100,53 @@ fn maybe_create_another_blocking_thread() {
...
+
+    // Calculate the amount of threads needed to spin up
+    // then retry sending while blocking. It doesn't spin if
+    // expected pool size is above the MAX_THREADS (which is a
+    // case won't happen)
+    let pool_size = EXPECTED_POOL_SIZE.load(Ordering::Relaxed);
+    let current_pool_size = CURRENT_POOL_SIZE.load(Ordering::SeqCst);
+    let reward = (AVR_FREQUENCY.load(Ordering::Relaxed) as f64 / 2.0_f64) as u64;
```
Comment (spacejam) [inline src/task/blocking.rs]: `Ordering::Acquire`

**Comment by spacejam** [comment]
In general, please change all of the atomics to use barriers that prevent potentially risky reorderings. Note that in the original, relaxed atomics were used with a clear explanation for why randomness in their use path violated no correctness criteria. In this code, it's much much much more complex, and relaxed atomics make this quite difficult to reason about. 

Mutexes will make this very easy for me to review and merge, and will not impact performance, so I recommend that you replace all of the atomic stuff with them, but I totally get that it's not as fun!

**Comment by spacejam** [comment]
After removing all Relaxed atomics, I have some concerns about long-term maintenance costs of this code due to the large amount of unserialized atomic access, and math without clear documentation. Please document the math more clearly for people like me who are less familiar with it, and please document the atomic access patterns with explanations of what the important correctness criteria are at each dependency site so that future contributors don't introduce bugs. 

Since bugs have been found in review, it would also make me feel better if some concurrency test was added that exercises the interleavings more.

You may think that this is a bit draconian, but lock-free programming is not easy for humans to get right, and this code is not obviously correct to me yet. 

This needs a bit more work before it's safe from a long-term maintenance perspective. The alternative is to wrap all atomic variables other than FREQUENCY in a Mutex so that their complex math is serialized. Please try this out if you're uninterested in adding concurrency tests. Mutex unlocks are less than 16ns, and as long as you're just doing this math on variables inside the Mutex and not doing any IO or running a blocking function with the Mutex held, it will be really really fast.

So, please make this code obviously correct by either using a simple Mutex around all of the variables other than FREQUENCY or by adding concurrent interleaving tests that will cause assumptions we're making to pop out.

File: benches/blocking.rs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+use async_std::task::blocking::JoinHandle;
+use futures::future::join_all;
+use std::thread;
+use std::time::Duration;
+use test::Bencher;
+
+#[bench]
+fn blocking(b: &mut Bencher) {
+    b.iter(|| {
+        let handles = (0..10_000)
```
Comment (kpp) [inline benches/blocking.rs]: That's great, how about swapning >MAX_THREADS with thread::sleep(10sec)? Will your patch change the behavior?

File: src/task/blocking.rs [hunk 10]
```diff
@@ -12,10 +13,29 @@ use lazy_static::lazy_static;
 use crate::future::Future;
 use crate::task::{Context, Poll};
 use crate::utils::abort_on_panic;
+use std::sync::{Arc, Mutex};
 
+// Low watermark value, defines the bare minimum of the pool.
+// Spawns initial thread set.
```
Comment (killercup) [inline src/task/blocking.rs]: even though the items aren't public these comments could all be doc comments

File: src/task/blocking.rs [hunk 11]
```diff
@@ -45,20 +78,84 @@ lazy_static! {
...
+}
+
+// Adaptive pool scaling function
+//
+// This allows to spawn new threads to make room for incoming task pressure.
+// Works in the background detached from the pool system and scales up the pool based
+// on the request rate.
+//
+// It uses frequency based calculation to define work. Utilizing average processing rate.
+fn scale_pool() {
```
Comment (killercup) [inline src/task/blocking.rs]: In another context there was some discussion around adding tracing (using log's `trace!` macro) to async-std. This functions seems like a good place for that, too.

Comment (killercup) [reply src/task/blocking.rs]: I just know that the idea was thrown around, at it was probably an opt-in feature, too. @stjepang or @yoshuawuyts would know more :)

**Comment by vertexclique** [comment]
So, to sum up, all the changes that I pushed:
* Values moved around are mutex rn.
* Now I cleanly calculate the values and also documented more explicitly how things are working. Even formula, very explicitly.
* Fixed #126 (OS-specific problem, hard to hit but possible)
* No significant degradation in performance. For the curious:
```
running 2 tests
test blocking        ... bench:  48,959,842 ns/iter (+/- 46,653,507)
test blocking_single ... bench:       3,322 ns/iter (+/- 8,452)
```
* I think I covered most of the cases with tests. Skipped all of them since they are exhaustive. But for the curious cats again here how you can run them:
```
$ cargo test longhauling_task_join -- --ignored --exact --nocapture
$ cargo test slow_join_interrupted -- --ignored --exact --nocapture
$ cargo test slow_join -- --ignored --exact --nocapture
```
* **NOTE**: Benchmarks cover the common cases. Tests are covering the possible edge cases.

File: src/task/blocking.rs [hunk 12]
```diff
@@ -1,5 +1,32 @@
...
+//!
+//! Algorithm is altered and adapted from [A Novel Predictive and Self–Adaptive Dynamic Thread Pool Management](https://doi.org/10.1109/ISPA.2011.61).
+//!
+//! ## Predictive Upscaler
+//! Selects upscaling amount based on estimation or when throughput hogs based on amount of tasks mapped.
+//! Throughput hogs determined by a combination of job in / job out frequency and current scheduler task assignment frequency.
+//!
+//! ## Time-based Downscaler
+//! After dynamic tasks spawned with upscaler they will continue working in between 1 second and 10 seconds.
+//! When tasks are detached from the channels after this amount they join back.
```
[the remark this answers is not in the document]
Comment (vertexclique) [reply src/task/blocking.rs]: You got totally right, my mistake of documenting. Will reword this.

File: src/task/blocking.rs [hunk 13]
```diff
@@ -45,20 +133,125 @@ lazy_static! {
...
         Pool { sender, receiver }
     };
+
+    /// Sliding window for pool task frequency calculation
+    static ref FREQ_QUEUE: Arc<Mutex<VecDeque<u64>>> = {
+        Arc::new(Mutex::new(VecDeque::with_capacity(FREQUENCY_QUEUE_SIZE + 1)))
+    };
+
+    /// Dynamic pool thread count variable
+    static ref POOL_SIZE: Arc<Mutex<u64>> = Arc::new(Mutex::new(LOW_WATERMARK));
```
Comment (spacejam) [inline src/task/blocking.rs]: All use of arcs in this module is unnecessary, please remove them all.

File: src/task/blocking.rs [hunk 14]
```diff
@@ -45,20 +133,125 @@ lazy_static! {
...
+///
+/// This allows to spawn new threads to make room for incoming task pressure.
+/// Works in the background detached from the pool system and scales up the pool based
+/// on the request rate.
+///
+/// It uses frequency based calculation to define work. Utilizing average processing rate.
+fn scale_pool() {
+    // Fetch current frequency, it does matter that operations are ordered in this approach.
+    let current_frequency = FREQUENCY.load(Ordering::SeqCst);
+    let freq_queue_arc = FREQ_QUEUE.clone();
```
Comment (spacejam) [inline src/task/blocking.rs]: there is no need to clone arcs unless you need another reference to the inner member. since there is a unique reference in the lazy_static, no arc clones, or arcs at all, are necessary.

File: src/task/blocking.rs [hunk 15]
```diff
@@ -45,20 +133,125 @@ lazy_static! {
...
+}
+
+/// Creates blocking thread to receive tasks
+/// Dynamic threads will terminate themselves if they don't
+/// receive any work after between one and ten seconds.
+fn create_blocking_thread() {
+    // Check that thread is spawnable.
+    // If it hits to the OS limits don't spawn it.
+    {
+        let current_arc = POOL_SIZE.clone();
```
Comment (spacejam) [inline src/task/blocking.rs]: no need to clone or to use an arc at all

File: src/task/blocking.rs [hunk 16]
```diff
@@ -68,29 +261,53 @@ fn maybe_create_another_blocking_thread() {
...
         .name("async-blocking-driver-dynamic".to_string())
         .spawn(move || {
             let wait_limit = Duration::from_millis(1000 + rand_sleep_ms);
 
-            DYNAMIC_THREAD_COUNT.fetch_add(1, Ordering::Relaxed);
+            // Adjust the pool size counter before and after spawn
+            {
+                let current_arc = POOL_SIZE.clone();
+                *current_arc.lock().unwrap() += 1;
+            }
```
Comment (spacejam) [inline src/task/blocking.rs]: cloning this arc does nothing, please remove the arc, its clone, and the subscope

File: src/task/blocking.rs [hunk 17]
```diff
@@ -68,29 +261,53 @@ fn maybe_create_another_blocking_thread() {
...
             while let Ok(task) = POOL.receiver.recv_timeout(wait_limit) {
                 abort_on_panic(|| task.run());
             }
-            DYNAMIC_THREAD_COUNT.fetch_sub(1, Ordering::Relaxed);
+            {
+                let current_arc = POOL_SIZE.clone();
+                *current_arc.lock().unwrap() -= 1;
+            }
```
Comment (spacejam) [inline src/task/blocking.rs]: cloning this arc does nothing, please remove the arc, its clone, and the subscope

File: src/task/blocking.rs [hunk 18]
```diff
@@ -68,29 +261,53 @@ fn maybe_create_another_blocking_thread() {
...
-        .expect("cannot start a dynamic thread driving blocking tasks");
+        .map_err(|err| {
+            match err.kind() {
+                ErrorKind::WouldBlock => {
+                    // Maximum allowed threads per process is varying from system to system.
+                    // Also, some systems have it(like macOS), and some don't(Linux).
+                    // This case expected not to happen.
+                    // But when happened this shouldn't throw a panic.
+                    let current_arc = POOL_SIZE.clone();
+                    MAX_THREADS.store(*current_arc.lock().unwrap() - 1, Ordering::SeqCst);
```
Comment (spacejam) [inline src/task/blocking.rs]: cloning this arc does nothing, please remove the arc, its clone, and the subscope

File: tests/thread_pool.rs [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+    let elapsed = start.elapsed().as_millis() - thread_join_time_max as u128;
+    println!(
+        "Long-hauling task join. Monotonic exec time: {:?} ns",
+        elapsed
+    );
+
+    // Should be less than 200_000 ns
+    // Previous implementation will panic when this test is running.
+    assert_eq!(elapsed < 200_000, true);
+}
```
Comment (spacejam) [inline tests/thread_pool.rs]: these tests are generally non-reproducible, and it's unclear if they are valuable. they tend to fail on my fairly modern 2018 i7 mbp

Comment (spacejam) [reply tests/thread_pool.rs]: when I said concurrency tests before, I meant adding randomized delays to operations that involve cross-thread coordination via atomics (or higher level primitives that invoke them internally). These might be OK as a benchmark so you can compare results on your own machine before and after changes, but I don't think they are useful as tests.

File: src/task/blocking.rs [hunk 20]
```diff
@@ -45,20 +133,125 @@ lazy_static! {
...
+/// Adaptive pool scaling function
+///
+/// This allows to spawn new threads to make room for incoming task pressure.
+/// Works in the background detached from the pool system and scales up the pool based
+/// on the request rate.
+///
+/// It uses frequency based calculation to define work. Utilizing average processing rate.
+fn scale_pool() {
+    // Fetch current frequency, it does matter that operations are ordered in this approach.
+    let current_frequency = FREQUENCY.load(Ordering::SeqCst);
```
Comment (spacejam) [inline src/task/blocking.rs]: It feels kind of weird that you're loading here and later storing 0, instead of swapping 0 in here and not ignoring items in the interval. would that be more correct?

File: src/task/blocking.rs [hunk 21]
```diff
@@ -10,21 +62,44 @@ use crossbeam_channel::{bounded, Receiver, Sender};
 use lazy_static::lazy_static;
 
 use crate::future::Future;
+use crate::io::ErrorKind;
 use crate::task::{Context, Poll};
 use crate::utils::abort_on_panic;
+use std::sync::{Arc, Mutex};
```
Comment (spacejam) [inline src/task/blocking.rs]: `Arc` unnecessary

File: tests/thread_pool.rs (hunk 19 above)
Comment (vertexclique) [reply tests/thread_pool.rs]: Please try running the tests with master's `blocking.rs`. They will panic.

File: src/task/blocking.rs (hunk 21 above)
Comment (vertexclique) [reply src/task/blocking.rs]: People tend to share this between threads. It is unnecessary because pool manager doesn't use it. If someone starts to share this across threads we will see the use of it.

File: src/task/blocking.rs (hunk 11 above)
Comment (yoshuawuyts) [reply src/task/blocking.rs]: `log` now has kv support, which we're already using in other places too. `tracing` can pick these up too, so using ` log` is probably the best choice here overall (:

Comment (davidbarsky) [reply src/task/blocking.rs]: I think we should get the `kv` support in `tracing-log` soon-ish? However I don't believe that `log` supports the `span`-like macros, so async-std might need to have its own, internal `span`-like macro and dispatch conditionally to async-log or tracing.

 I haven't considered how the differences between [async-log's `span`](https://docs.rs/async-log/2.0.0/async_log/macro.span.html) and [`tracing`'s span](https://docs.rs/tracing/0.1.7/tracing/macro.span.html) can be bridged, but I think it's feasible if `tracing'`s `span!().enter()` is used to mimic async-log's nested `span` block.

Comment (yoshuawuyts) [reply src/task/blocking.rs]: @davidbarsky we could use [`async-log::span`](https://docs.rs/async-log/2.0.0/async_log/macro.span.html) for this, probably. Though there's still a few open issues on the repo, the interface wouldn't change.

Comment (davidbarsky) [reply src/task/blocking.rs]: @yoshuawuyts Yep! I think we're on the same page about wanting to bridge the two libraries. 

I'm spitballing _one_ mechanism on bridging that gap, which consists of using mutually exclusive feature flags in a `build.rs` to dispatch to either `async-log` or `tracing`. This would entail using an `async-std`-specific `span` macro until the gaps—if any exist—between `async-std`'s `span` and `tracing`'s `span` are closed.

I'm sorry if what I communicated wasn't clear!

File: tests/thread_pool.rs (hunk 19 above)
Comment (spacejam) [reply tests/thread_pool.rs]: They panic on this branch on a fairly modern machine, so I'm not sure they are useful on machines other than yours. In general, timing-based performance tests are an antipattern, as they have big reproducibility issues. If you want to do this kind of testing, it's better to use something like cachegrind to count CPU instructions executed.

File: src/task/blocking.rs [hunk 22]
```diff
@@ -45,52 +133,175 @@ lazy_static! {
...
+/// +--------+-----------------+----------------------------------+
+/// ```
+/// Under these definitions formula is following:
+/// ```text
+/// EMA = α * [ Yt + (1 - α)*Yt-1 + ((1 - α)^2)*Yt-2 + ((1 - α)^3)*Yt-3 ... ] + St
+/// ```
+/// # Arguments
+///
+/// * `freq_queue` - Sliding window of frequency samples
+#[inline]
```
Comment (spacejam) [inline src/task/blocking.rs]: in general, humans are bad at guessing whether something should be inlined. Usually in rust it's better to leave these attributes out unless come code is going to be executed in a super hot path and you've done work to ensure the improvement is measurable

File: tests/thread_pool.rs (hunk 19 above)
Comment (spacejam) [reply tests/thread_pool.rs]: Because they don't actually assert on anything now (which was what caused them to fail on my laptop), I think they are better as benchmarks, right? I don't think it makes sense to have tests that can't fail.

File: src/task/blocking.rs (hunk 22 above)
Comment (vertexclique) [reply src/task/blocking.rs]: This comment is not contributing to the overall discussion + review. I’ve found it quite a bit `out of context` and also from the `book`. I've made the benchmarks already with and without and decided to leave it inlined. You can see below my last run again after your comment (run shouldn't be needed because of how compiler inlines and unrolls):

Mean of 10 consecutive runs of `blocking` benchmark:
```
43057207.9 ns without inlining
41816626.1 ns with inlining
```

That is also better for users of this library indirectly.

File: tests/thread_pool.rs (hunk 19 above)
Comment (vertexclique) [reply tests/thread_pool.rs]: > I don't think it makes sense to have tests that can't fail.

Please read my comments carefully. You can see what it tests in action:
https://asciinema.org/a/R6b3UcFo0xfNo1sCPOPJrB04T

> when I said concurrency tests before, I meant adding randomized delays to operations that involve cross-thread coordination via atomics (or higher level primitives that invoke them internally).

Unexposed architecture shouldn't be tested with whitebox fashion.

Comment (spacejam) [reply tests/thread_pool.rs]: Where are the assertions that fail? The asciicinema is too long to watch. 

Possible concurrent interleavings are exposed architecture, and this kind of test has resulted in a huge number of bugs being discovered in various lock-free projects I've been involved in. Concurrency is really hard to get right sometimes, and the original PR with more unserialized atomic accesses would have had bugs jump out pretty quickly with this approach.

File: src/task/blocking.rs (hunk 22 above)
Comment (spacejam) [reply src/task/blocking.rs]: I'm a bit skeptical that change is related to the inline here because this is a function that gets called 5 times per second, and that measurement skew is 2 milliseconds.

**Comment by vertexclique** [comment]
Please track the progress on https://github.com/async-rs/async-std/pull/181
